alibi_detect/cd/keops/learned_kernel.py

```diff
@@ -0,0 +1,341 @@
+        dataset
+            Dataset object used during training.
+        dataloader
+            Dataloader object used during training. Only relevant for 'pytorch' backend.
+        input_shape
+            Shape of input data.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+        super().__init__(
+            x_ref=x_ref,
+            p_val=p_val,
+            x_ref_preprocessed=x_ref_preprocessed,
+            preprocess_at_init=preprocess_at_init,
+            update_x_ref=update_x_ref,
+            preprocess_fn=preprocess_fn,
+            n_permutations=n_permutations,
+            train_size=train_size,
+            retrain_from_scratch=retrain_from_scratch,
+            input_shape=input_shape,
+            data_type=data_type
+        )
+        self.meta.update({'backend': 'keops'})
+
+        # Set device, define model and training kwargs
+        self.device = get_device(device)
+        self.original_kernel = kernel
+        self.kernel = deepcopy(kernel)
+
+        # Check kernel format
+        self.has_proj = hasattr(self.kernel, 'proj') and isinstance(self.kernel.proj, nn.Module)
+        self.has_kernel_b = hasattr(self.kernel, 'kernel_b') and isinstance(self.kernel.kernel_b, nn.Module)
+
+        # Define kwargs for dataloader and trainer
+        self.dataset = dataset
+        self.dataloader = partial(dataloader, batch_size=batch_size, shuffle=True,
+                                  drop_last=True, num_workers=num_workers)
+        self.train_kwargs = {'optimizer': optimizer, 'epochs': epochs,  'preprocess_fn': preprocess_batch_fn,
+                             'reg_loss_fn': reg_loss_fn, 'learning_rate': learning_rate, 'verbose': verbose}
+        if isinstance(train_kwargs, dict):
+            self.train_kwargs.update(train_kwargs)
+
+        self.j_hat = LearnedKernelDriftKeops.JHat(
+            self.kernel, var_reg, self.has_proj, self.has_kernel_b).to(self.device)
+
+        # Set prediction and permutation batch sizes
+        self.batch_size_predict = batch_size_predict
+        self.batch_size_perms = batch_size_permutations
+        self.n_batches = 1 + (n_permutations - 1) // batch_size_permutations
+
+    class JHat(nn.Module):
+        """
+        A module that wraps around the kernel. When passed a batch of reference and batch of test
+        instances it returns an estimate of a correlate of test power.
+        Equation 4 of https://arxiv.org/abs/2002.09116
+        """
+        def __init__(self, kernel: nn.Module, var_reg: float, has_proj: bool, has_kernel_b: bool):
+            super().__init__()
+            self.kernel = kernel
+            self.has_proj = has_proj
+            self.has_kernel_b = has_kernel_b
+            self.var_reg = var_reg
+
+        def forward(self, x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+            n = len(x)
+            if self.has_proj:
```

Comment:
It might be worth adding a comment on why we have to have this `type: ignore`? Since we can probably remove it if we make the kernel definitions more strict later on...

Reply:
Should be fixed now. `self.proj` is actually a `nn.Module` so the updated if statement simply checks for that.